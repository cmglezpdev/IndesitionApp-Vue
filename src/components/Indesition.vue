<template>
    <img v-if="img" v-bind:src="img" alt="bg">
    <div class="bg-dark"></div>

    <div class="indesition-container">
        <input v-model="question" type="text" placeholder="Hazme una pregunta"> 
        <p>Recuerda terminar con un signo de interrogacion (?)</p>
        <div v-if="isValidCuestion">
            <h2>{{ question }}</h2>
            <h1>{{ answer }}</h1>
        </div>
    </div>

</template>

<script>
export default {
    data() {
        return {
            question: null,
            answer: null,
            img: null,
            isValidCuestion: false
        }
    },
    methods: {

        async getAnswer() {

            this.answer = "Pensando ...";
            const url = "https://yesno.wtf/api";
        
            const { answer, image } = await fetch(url).then( result => result.json() )

            switch( answer ) {
                case "no":
                    this.answer = "NO!!";
                    break;
                case "yes":
                    this.answer = "SII!!"
                    break;
                case "mybe":
                    this.answer = "QUISAS!!"
                    break;
            }

            this.img = image;

        }

    },
    watch: {
        question( value, oldValue ) {
            this.isValidCuestion = false;
            if( !value.includes('?') ) return 
            this.isValidCuestion = true;
            this.getAnswer();
        }
    }
}
</script>


<style>

img, .bg-dark {
    height: 100vh;
    left: 0;
    max-width: 100%;
    max-height: 100%;
    position: fixed;
    top: 0;
    width: 100vw;
}

.bg-dark {
    background-color: rgba(0, 0, 0, 0.4);
}

.indesition-container {
    position: relative;
    z-index: 10;
}

input {
    width: 350px;
    padding: 10px 15px;
    border-radius: 5px;
    border: none;
}
input:focus {
    outline: none;
}

p {
    color: white;
    font-size: 20px;
    margin-top: 5px;
}

h1, h2 {
    color: white;
}

h2 {
    margin-top: 150px;
}
</style>