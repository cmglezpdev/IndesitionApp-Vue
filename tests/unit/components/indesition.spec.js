import { shallowMount } from '@vue/test-utils';
import Indesition from '@/components/Indesition.vue';

describe("Indesition Component", () => {

    let wrapper;
    // Mocks de un fetch Api
    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            answer: 'yes',
            forced: false,
            image: 'https://yesno.wtf/assets/yes/2.gif'
        })
    }));


    beforeEach(() => {
        wrapper = shallowMount( Indesition );
        jest.clearAllMocks();
    })

    test('debe de hacer match con el SnapShot', () => {
        expect( wrapper.html()).toMatchSnapshot();
    });

    test('Escribir en el input no debe de disparar nada', async () => {
        
        const getAnswerSpy = jest.spyOn( wrapper.vm, "getAnswer" );
        
        const input = wrapper.find('input');
        await input.setValue('Hola Mundo');

        expect( getAnswerSpy ).not.toHaveBeenCalled();
    });

    test('Al escribir simbolo "?" debe de disparar el getAnswer', async () => {
        
        const getAnswerSpy = jest.spyOn( wrapper.vm, "getAnswer" );
        
        const input = wrapper.find('input');
        await input.setValue('Hola Mundo?');

        expect( getAnswerSpy ).toHaveBeenCalled();

    });

    test('Pruebas en getAnswer', async () => {
        
        await wrapper.vm.getAnswer();
    
        const img = wrapper.find("img");
        expect(img.exists()).toBeTruthy();
        expect(wrapper.vm.img).toBe("https://yesno.wtf/assets/yes/2.gif");
        expect(wrapper.vm.answer).toBe('SII!!');
    });

    test('pruebas en getAnswer - fallo en el API', async () => {
        
        fetch.mockImplementationOnce(() => Promise.reject('Api is down'));

        await wrapper.vm.getAnswer();

        const img = wrapper.find("img");
        expect(img.exists()).toBeFalsy();
        expect(wrapper.vm.answer).toBe('No se pudo cargar del Api');
    });


})