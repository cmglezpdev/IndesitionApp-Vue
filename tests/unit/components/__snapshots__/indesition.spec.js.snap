// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indesition Component debe de hacer match con el SnapShot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indesition-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con un signo de interrogacion (?)</p>
  <!--v-if-->
</div>
`;
